lib/src/wallet/keys/unified.rs

```diff
@@ -0,0 +1,406 @@
+    }
+}
+
+#[test]
+fn read_write_receiver_selections() {
+    for (i, receivers_selected) in (0..8)
+        .into_iter()
+        .map(|n| ReceiverSelection::read([1, n].as_slice(), ()).unwrap())
+        .enumerate()
+    {
+        let mut receivers_selected_bytes = [0; 2];
+        receivers_selected
+            .write(receivers_selected_bytes.as_mut_slice())
+            .unwrap();
+        assert_eq!(i as u8, receivers_selected_bytes[1]);
+    }
+}
+
+impl UnifiedSpendAuthority {
+    pub fn addresses(&self) -> &[UnifiedAddress] {
+        &self.addresses
+    }
+
+    pub fn transparent_child_keys(&self) -> &Vec<(usize, secp256k1::SecretKey)> {
+        &self.transparent_child_keys
+    }
+
+    pub fn new_address(
+        &mut self,
+        desired_receivers: ReceiverSelection,
+    ) -> Result<UnifiedAddress, String> {
+        let orchard_receiver = if desired_receivers.orchard {
+            Some(
+                orchard::keys::FullViewingKey::from(&self.orchard_key)
+                    .address_at(self.addresses.len(), Scope::External),
+            )
+        } else {
+            None
+        };
+        let sapling_receiver = if desired_receivers.sapling {
+            let (mut new_index, address) =
+                zcash_primitives::zip32::ExtendedFullViewingKey::from(&self.sapling_key)
+                    .find_address(self.next_sapling_diversifier_index)
+                    .expect("Diversifier index overflow");
+            new_index.increment().expect("Diversifier index overflow");
+            self.next_sapling_diversifier_index = new_index;
+            Some(address)
+        } else {
+            None
+        };
+        let transparent_receiver = if desired_receivers.transparent {
+            let key_index =
+                KeyIndex::hardened_from_normalize_index(self.addresses.len() as u32).unwrap();
+            let new_key = self
+                .transparent_parent_key
+                .derive_private_key(key_index)
+                .expect("Private key derevation failed")
+                .private_key;
+            let secp = secp256k1::Secp256k1::new();
+            let address = secp256k1::PublicKey::from_secret_key(&secp, &new_key);
+            self.transparent_child_keys
+                .push((self.addresses.len(), new_key));
+            Some(address)
+        } else {
+            None
+        };
+        let ua = UnifiedAddress::from_receivers(
+            orchard_receiver,
+            sapling_receiver,
+            #[allow(deprecated)]
+            transparent_receiver
+                .as_ref()
+                // This is deprecated. Not sure what the alternative is,
+                // other than implementing it ourselves.
+                .map(zcash_primitives::legacy::keys::pubkey_to_address),
+        )
+        .ok_or(format!(
+            "Invalid receivers requested! At least one of sapling or orchard required"
+        ))?;
+        self.addresses.push(ua.clone());
+        Ok(ua)
```

Comment:
Instead of mutating the USA in place, and also returning the same UA, why not enforce a single interface on the USA that provides the UA to interested callers?

Reply:
We could. I thought it would be convenient to return the address of the newly generated UA.

Reply:
It _is_ convenient, but there's a clarity cost.   Your call.